### PR TITLE
Clean Navigation Layout

### PR DESCRIPTION
**Pull request by lapp0**
- Use navbar and **one** attribute list for navigation (fixes https://github.com/nix-gui/nix-gui/issues/70)
- Implement skeleton of `ListOf` (obviates https://github.com/nix-gui/nix-gui/pull/76)
  - https://github.com/nix-gui/nix-gui/issues/30 still incomplete, but with progess

`origin/master`
image

`origin/clean-navigation`
image

fixes https://github.com/nix-gui/nix-gui/issues/31 (not really, it actually just deletes `StringListField` and assumes `DynamicListOf[TextField]` will handle it

### CODE REVIEW

**Comment by adrianparvino** [comment]
General look and field comment for now: I think having 3-4 (maybe dependent on the window width or user-customizable?) lists makes sense. Or maybe even make it anchor around `attrsOf` or `listOf`? I'm for the navigation bar, but having only 2 columns feels too cramped.

**Comment by lapp0** [comment]
I disagree, but I could be wrong. We should leave this as a question for users who volunteer to UX test.

https://github.com/nix-gui/nix-gui/issues/87

File: nixui/graphics/field_widgets.py [hunk 1]
```diff
@@ -404,10 +372,9 @@ def validate_field(self, value):
         return value in self.choices
 
     def load_value(self, value):
-        if not self.validate_field(value):
-            value = ''
-        self.setCurrentIndex(self.choices.index(value))
-        self.loaded_value = value
+        if self.validate_field(value):
```
Comment (adrianparvino) [inline nixui/graphics/field_widgets.py:375]: It seems to only be changed for this specifically, should this be done for all?

File: nixui/graphics/navbar.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, option_path, set_option_fn, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.set_option_fn = set_option_fn
+
+        # create widgets and define behavior
+        back_btn = QtWidgets.QPushButton('◀')
+        back_btn.clicked.connect(lambda: print('not implemented'))
+
+        up_btn = QtWidgets.QPushButton('▲')
```
Comment (adrianparvino) [inline nixui/graphics/navbar.py:46]: Not so sure if we should have a back button if we already have breadcrumbs. I think an Clicked on the breadcrumb component makes more sense.

File: nixui/options/attribute.py [hunk 3]
```diff
@@ -37,7 +38,9 @@ def __lt__(self, other):
 
     def __str__(self):
         return '.'.join([
-            f'"{attribute}"' if '.' in attribute else attribute
+            attribute
+            if re.match(r'^[a-zA-Z\_][a-zA-Z0-9\_\'\-]*$', attribute)
```
Comment (adrianparvino) [inline nixui/options/attribute.py:42]: I'm not sure if this regex should include `\'`.

**Review by adrianparvino — COMMENTED** [review]
Here's my technical review:

File: nixui/graphics/navbar.py (hunk 2 above)
Comment (lapp0) [reply nixui/graphics/navbar.py:46]: We don't have breadcrumbs right now unfortunately. I'll make an issue for that though. 

https://github.com/nix-gui/nix-gui/issues/88

File: nixui/options/attribute.py (hunk 3 above)
Comment (lapp0) [reply nixui/options/attribute.py:42]: Hmm, it's based on https://github.com/NixOS/nix/blob/master/src/libexpr/lexer.l#L97

File: nixui/graphics/field_widgets.py (hunk 1 above)
Comment (lapp0) [reply nixui/graphics/field_widgets.py:375]: I believe so. If there isn't a valid choice for one-of, we don't set anything. This change isn't really part of the goal of this PR, but a bug I found.

File: nixui/graphics/navbar.py (hunk 2 above)
Comment (lapp0) [reply nixui/graphics/navbar.py:46]: Additionally, the back arrow should take us to the previous view, be it "search" or "options" (tree). This isn't always in the breadcrumbs.

Comment (adrianparvino) [reply nixui/graphics/navbar.py:46]: Right, my bad. I meant to say the "up one level" button.

File: nixui/options/attribute.py (hunk 3 above)
Comment (adrianparvino) [reply nixui/options/attribute.py:42]: Thanks. I think it's worth it to add a comment here, referencing either your link or https://github.com/NixOS/nixpkgs/blob/8da27ef161e8bd0403c8f9ae030ef1e91cb6c475/pkgs/tools/nix/nixos-option/libnix-copy-paste.cc#L52

File: nixui/graphics/navbar.py (hunk 2 above)
Comment (lapp0) [reply nixui/graphics/navbar.py:46]: ok, resolving in favor of #88

File: nixui/options/attribute.py (hunk 3 above)
Comment (lapp0) [reply nixui/options/attribute.py:47]: sounds good

resolved with https://github.com/nix-gui/nix-gui/pull/80/commits/82b5b451cc3b3d6a1d1d73553785978a8129b5cb

File: nixui/graphics/field_widgets.py (hunk 1 above)
Comment (adrianparvino) [reply nixui/graphics/field_widgets.py:375]: I think we should make another issue for this and have just a general bug squashing PR.

Comment (lapp0) [reply nixui/graphics/field_widgets.py:375]: Sure, but we should include this one here for ease rather than reverting IMO.

Comment (adrianparvino) [reply nixui/graphics/field_widgets.py:375]: Yep. We should keep this one.

**Review by adrianparvino — APPROVED** [review]

**Pull request merged**